### PR TITLE
UCT/API: Added uct_iface params structure

### PR DESCRIPTION
**Pull request by brminich**
uct_iface_open is modified to accept uct_iface_params_t structure as a parameter.
This should be useful for various extensions (tag matching, connection management, etc). New parameters will be added as additional fields of this structure.

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2459/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1801/ for details (Mellanox internal link).

**Comment by brminich** [comment]
@MattBBaker, could you please post ORNL failure?

**Comment by shamisp** [comment]
@MattBBaker this one may require a test on Cray.

**Comment by MattBBaker** [comment]
Build error.
```
  CXX    uct/gtest-test_stats.o
/var/lib/jenkins/jobs/UCX/workspace/contrib/../test/gtest/uct/uct_test.cc: In constructor ‘uct_test::entity::entity(const resource&, uct_iface_config_t*, size_t, uct_md_config_t*)’:
/var/lib/jenkins/jobs/UCX/workspace/contrib/../test/gtest/uct/uct_test.cc:212: error: expected primary-expression before ‘.’ token
/var/lib/jenkins/jobs/UCX/workspace/contrib/../test/gtest/uct/uct_test.cc:213: error: expected primary-expression before ‘.’ token
/var/lib/jenkins/jobs/UCX/workspace/contrib/../test/gtest/uct/uct_test.cc:214: error: expected primary-expression before ‘.’ token
make[2]: *** [uct/gtest-uct_test.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: Leaving directory `/var/lib/jenkins/jobs/UCX/workspace/build-tests/test/gtest'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/var/lib/jenkins/jobs/UCX/workspace/build-tests'
make: *** [all] Error 2
Build step 'Execute shell' marked build as failure
```

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2463/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1804/ for details (Mellanox internal link).

**Comment by brminich** [comment]
@MattBBaker, is the fault different now?

**Comment by MattBBaker** [comment]
@brminich It is, but I can't see what actually went wrong. It looks like it just terminated. I'll have it retest.

**Comment by MattBBaker** [comment]
bot:ornl:retest

File: src/uct/api/uct.h [hunk 1]
```diff
@@ -296,6 +296,18 @@ struct uct_iface_attr {
 
 
 /**
+ * @ingroup UCT_RESOURCE
+ * @brief Parameters used for interface creation.
+ */
+struct uct_iface_params {
+    char                     *tl_name;    /**< Transport name */
```
Comment (yosefe) [inline src/uct/api/uct.h]: need to be const char *

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2466/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1807/ for details (Mellanox internal link).

**Comment by brminich** [comment]
resolved merge conflicts

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1808/ for details (Mellanox internal link).

**Comment by brminich** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2470/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2471/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1811/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
:+1: 
@shamisp pls take a look

File: src/uct/api/uct.h [hunk 2]
```diff
@@ -311,6 +311,18 @@ struct uct_iface_attr {
 
 
 /**
+ * @ingroup UCT_RESOURCE
+ * @brief Parameters used for interface creation.
```
Comment (shamisp) [inline src/uct/api/uct.h]: besides brief, please add full doc for the structure

Comment (brminich) [inline src/uct/api/uct.h]: @shamisp, please check the update

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2481/ for details.

**Comment by yosefe** [comment]
:+1:

**Comment by yosefe** [comment]
:+1:

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1825/ for details (Mellanox internal link).

**Pull request merged**